### ISSUE TITLE
Upgrade Guide for v2

### ISSUE THREAD

**Issue by martincostello** [microsoft/OpenAPI.NET#1973]
Howdy folks - after a pointer from @captainsafia, I've just been having a quick look at what would happen if I tried to use Microsoft.OpenApi 2.0.0 in [Swashbuckle.AspNetCore](https://github.com/domaindrivendev/Swashbuckle.AspNetCore) using the latest preview.

Support for OpenAPI 3.1 has been a populate ask from our users for quite some time: https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/2349.

Unsurprisingly, there's quite a lot of breaking changes to deal with:

![Image](https://github.com/user-attachments/assets/54ad10d5-7de8-4bf4-a798-2639fac16137)

Rather than me just blindly chip through those changes until it compiles (let alone tests pass), I figured a migration guide would be a useful starting point. Even if it's just a simple list of changes rather than a full-blown document, as clearly it would be a moving goal-post in the middle of your development, that would go a good way to understanding the effort that would be required for us in more detail.

This looks like it's going to be a non-trivial piece of work to adopt, and will put constraints on us/our users on what functionality will and will not be retained moving forwards compared to v1 and OpenAPI 2.0/3.0.

Things I've found so far:

- Removal of the JSON reader;
- Parsing documents is async;
- Removal of `IOpenApiAny`;
- `JsonSchemaType` replacing `string` for schema types;
- `OpenApiDocument.ResolveReference()` being removed.

Would someone be able to put something together to aid us in preparing to consume version 2? Ideally this is something we (Swashbuckle) can work on in the background so that we could land support for 3.1 relatively soon after you ship 2.0.0 as stable.

**Comment by RachitMalik12** [comment]
@martincostello We also have some release notes here: https://github.com/microsoft/OpenAPI.NET/releases/tag/2.0.0-preview1 that contain details on the breaking changes. Feel free to take a look and see if it helps you get started while we work on the blogpost.

**Comment by baywet** [comment]
Thanks for the nudge here @martincostello. We've already started working on an internal document for the blog post for the announcement of the release which contains a lot of the information you'd be looking for in an upgrade scenario.

I've asked @RachitMalik12 our dear PM to abstract the key information in a markdown document on this repo in a document [similar to this one](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/main/docs/upgrade-to-v5.md)

**Comment by martincostello** [comment]
Any updates here? I had been using the release notes for the 2.0.0-preview.1 release in GitHub, but it's been removed 😅.

**Comment by martincostello** [comment]
Yep, sorry my bad, it had flipped over onto page 2, but I thought 1 and 2 had disappeared as it jumped down to 1.6.23 and when I tried to open it by manually tweaking the URL I got a 404 (the latest release has a dot between preview and the number where it didn't before):

![Image](https://github.com/user-attachments/assets/26052713-088e-413f-a575-214b29aa8f94)

**Comment by martincostello** [comment]
I'm trying to update Swashbuckle.AspNetCore from 2.0.0-preview5 to 2.0.0-preview7 and it's proving _incredibly_ difficult due to the huge breaking changes between the two versions.

Interfaces aren't always useable as they're a read-only model (e.g. `Example` only has a `get`) which means needing to use `OpenApiSchema` or `OpenApiSchemaReference`, but it isn't clear what to do in the scenario where the interface isn't either of these concrete types. It's also not clear where changes are intentional or are maybe a bug/missing feature.

Even a work-in-progress upgrade guide would be useful to refer to to try and get things working as this is shaping up to be quite a large piece of work to migrate to. I'd prefer to get ahead of the game and help find bugs and give feedback (e.g. #2253) well in advance of 2.0.0 being "done".

**Comment by baywet** [comment]
Rachit has started a draft PR, but on this repo... which is private https://github.com/MicrosoftDocs/openapi-docs-pr/pull/192 (don't ask me why we need a private repo + a public fork, it's a requirements from our friends in the docs team...)

If that can help, I can share a PDF printout of what we currently have, let me know.

As for the interfaces change, the idea is that consumers like you will use type assertions, in later previews we've also added additional properties that give you the target recursively as well. The main issue we were trying to fix (beyond having a working implementation of references) was to let users know when they might actually be mutating the target when they thought they were mutating a local reference/copy instead.

So a couple of examples:

```csharp

IOpenApiSchema someSchema = // does not matter how you got it;

if (someSchema is OpenApiSchemaReference schemaReference)
{
    // it's a reference, just read things
    var schemaType = schemaReference.Type;
    // or update the description/summary for the reference
    schemaReference.Description = "bar";
    // or obtain the target
    IOpenApiSchema target = schemaReference.Target;
    OpenApiSchema definitiveTarget = schemaReference.RecursiveTarget;
}
else if (someSchema is OpenApiSchema concreteSchema)
{
    // it's a schema directly, do whatever you want to it
}
```

Let me know if you have any additional comments or questions.

**Comment by martincostello** [comment]
A PDF of the current state would be useful, thanks.

Regarding interfaces vs. types, my main question would be about the `else`.

What would one do in the case that the interface isn't either of those two implementations but something else?

**Comment by baywet** [comment]
I'd say throw because something went wrong. We only have two implementations of each of those interfaces. I don't see any reason why we'd have more at this point. Even if the reference needs to be specialized, the specialized type will derive from reference.

The only reason why there might be additional implementations is if the consumer brings additional implementations, again that's possible, but I'm not sure why would anybody do that?

**Comment by baywet** [comment]
There is the export, some of the reference stuff in there is NOT up to date with the code, which is why we haven't released it yet. Let me know if the formatting is too painful, I'll see what I can do.

[upgrade-guide-oai-2.pdf](https://github.com/user-attachments/files/19390104/upgrade-guide-oai-2.pdf)

**Comment by martincostello** [comment]
Thanks!

**Comment by Kirkmasterix** [comment]
Hello, i have a problem with this:
builder.Services.AddSwaggerGen(c =>
{
    c.AddSecurityDefinition("bearerAuth", new OpenApiSecurityScheme
    {
        Name = "Authorization",
        Type = SecuritySchemeType.Http,
        Scheme = "Bearer",
        BearerFormat = "JWT",
        In = ParameterLocation.Header,
        Description = "JWT Authorization header using the Bearer scheme."
    });
    c.AddSecurityRequirement(new OpenApiSecurityRequirement
    {
        {
            new OpenApiSecurityScheme
            {
                Reference = new OpenApiReference
                {
                    Type = ReferenceType.SecurityScheme,
                    Id = "Bearer"
                }
            },
            Array.Empty<string>() 
        }
    });
});
Reference was removed from OpenApiSecurityScheme in net 10.0 and I cant find any solution

**Comment by martincostello** [comment]
See [Add Security Definitions and Requirements](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/v10.0.0/docs/configure-and-customize-swaggergen.md#add-security-definitions-and-requirements)

**Comment by Kirkmasterix** [comment]
Thanks Martin
But there is something that I cant understand, I am trying to add the swagger to my services in a webapp in .net 10.0
But now that I apply the code you sent me I cant enter it a Bearer token, it doesnt authroized me, I dont understand how to modify the new code to get a swagger with AddSecurityDefinition with bearer token and global AddSecurityRequirement to prtect my endpoints

**Comment by martincostello** [comment]
@Kirkmasterix Take a look at the document I linked you to and compare the code examples to your own code and see how the definition syntax has changed.
